### PR TITLE
Enable minion's IPC channel to aggregate results from spawned jobber processes

### LINKED ISSUE #61274: [FEATURE REQUEST] Salt Minion should cache ReqChannels.
**Is your feature request related to a problem? Please describe.**

Looks like there may be a bottleneck in minion.py: send_req_sync() and minion.py:send_req_async() below, where we are re-creating channels (instead of caching them). At scale (about 5K minions), this causes a lot of heavy channel creation operations when minions respond and the delay is noticeable.

I've only tested with rabbitmq, but the same issue may exist for zeromq.

**Describe the solution you'd like**

If possible cache a single request channel for each minion.

### PR DESCRIPTION
**Pull request by devkits**
### What does this PR do?

Enable minion's IPC channel to aggregate results from spawned jobber processes. Use a long-running request channel in the minion parent process to communicate job results back to the master via broker-based or broker-less transport.

This is a necessary optimization at scale for transports that prefer a sustained long-running connection because connection create/dispose operations are expensive. The working assumption is that this change benefits all supported transports.

Testing Done:
 * this tests provide coverage for this use case:
      .../salt/tests/pytests/integration/minion.*

### Previous Behavior
Minion would spawn a process that would run a job and create/dispose a new transport connection just to communicate one job result. This does not scale well for transports that prefer persistent connections because connection operations are expensive.

### New Behavior
Minion would spawn a process that would run a job and communicate result via an IPC channel to the parent minion process. Parent minion process would use a long-running connection to communicate result back to master. This approach scales better because it reduces connection churn when thousands of minions are responding to the master. 

Fixes: #61274

### CODE REVIEW

**Comment by devkits** [comment]
re-run full all

File: salt/minion.py [hunk 1]
```diff
@@ -1364,11 +1364,41 @@ def connect_master(self, failed=False):
         """
         Return a future which will complete when you are connected to a master
         """
+        # Consider refactoring so that eval_master does not have a subtle side-effect on the contents of the opts array
```
Comment (jfindlay) [inline salt/minion.py:1367]: Someday, `opts` (`__opts__`, `__salt__`, *et al*) will be readonly, rite?

**Comment by jfindlay** [comment]
You might consider also updating https://github.com/saltstack/salt/blob/master/doc/topics/development/architecture.rst#minion-job-flow.

**Comment by devkits** [comment]
> You might consider also updating https://github.com/saltstack/salt/blob/master/doc/topics/development/architecture.rst#minion-job-flow.

Thanks. I will update https://github.com/saltstack/salt/blob/master/doc/topics/development/architecture.rst#minion-job-flow shortly.

**Comment by devkits** [comment]
re-run full all

**Comment by dwoz** [comment]
@devkits [This test](https://jenkins.saltproject.io/job/pr-centos-7-x86_64-py3-tcp-pytest/job/PR-61468/8/testReport/tests.pytests.integration.minion/test_return_retries/test_publish_retry/) needs to be fixed. Otherwise, the PR looks good.

**Comment by devkits** [comment]
> 

Thanks. I can reproduce the test failure locally. Looks like the fix may need  be in the TCP transport itself. I will try to update the PR with the fix.

**Review by dwoz — DISMISSED** [review]

**Review by twangboy — DISMISSED** [review]

File: salt/minion.py (hunk 1 above)
Comment (dwoz) [reply salt/minion.py:1368]: I'm not sure how quickly that could happen. In any case, we want to tamper with `__opts__` as little as possible.

**Review by s0undt3ch — DISMISSED** [review]

**Comment by garethgreenaway** [comment]
@devkits Can you please add a changelog for this PR?  Thanks!

**Pull request merged**